### PR TITLE
refactor: remove @dcl/crdt & unify CRDT logic inside components

### PR DESCRIPTION
**Pull request by menduz**
This is the preface to implement the APPEND command for the CRDT. The PR mainly removes the double book-keeping of the CRDTs in favor of keeping the whole control of the state to the Component itself.

It greatly improves memory usage and executed operations per frame. Without losing any functionality.

The only loss is that the state of unknown components will no-longer be stored. But the messages will still be valid and broadcasted to other actors of the system that may use them.

The CRDT logic is now encapsulated for components in one convenience helper. You can find the tests at `test/ecs/crdt-rules.spec.ts`

### CODE REVIEW

File: test/ecs/crdt-rules.spec.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
+import { CrdtMessageType, Entity } from '../../packages/@dcl/ecs/src'
+import { createUpdateFromCrdt } from '../../packages/@dcl/ecs/src/engine/component'
+import { ByteBuffer } from '../../packages/@dcl/ecs/src/serialization/ByteBuffer'
+
+describe('Conflict resolution rules for LWW-ElementSet based components', () => {
```
Comment (menduz) [inline test/ecs/crdt-rules.spec.ts:5]: Here are the new tests

File: test/snapshots/billboard.ts.crdt [hunk 2]
```diff
@@ -48,31 +48,31 @@ CALL onUpdate(0.0)
   CRDT: e=0x203 c=1018 t=1 data={"mesh":{"$case":"plane","plane":{"uvs":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,0,0]}}}
   CRDT: e=0x204 c=1018 t=1 data={"mesh":{"$case":"plane","plane":{"uvs":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,0,0]}}}
   CRDT: e=0x205 c=1018 t=1 data={"mesh":{"$case":"plane","plane":{"uvs":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,0,0]}}}
-  OPCODES ~= 189k
-  MALLOC_COUNT = 450
-  ALIVE_OBJS_DELTA ~= 0.15k
+  OPCODES ~= 107k
+  MALLOC_COUNT = 63
```
Comment (menduz) [inline test/snapshots/billboard.ts.crdt:52]: Performance impact

File: test/snapshots/cubes.ts.crdt [hunk 3]
```diff
@@ -561,9 +561,9 @@ CALL onUpdate(0.0)
   CRDT: e=0x2b4 c=1018 t=1 data={"mesh":{"$case":"box","box":{"uvs":[]}}}
   CRDT: e=0x2b5 c=1018 t=1 data={"mesh":{"$case":"box","box":{"uvs":[]}}}
   CRDT: e=0x2b6 c=1018 t=1 data={"mesh":{"$case":"box","box":{"uvs":[]}}}
-  OPCODES ~= 1443k
-  MALLOC_COUNT = 6812
-  ALIVE_OBJS_DELTA ~= 2.39k
+  OPCODES ~= 904k
+  MALLOC_COUNT = 1846
+  ALIVE_OBJS_DELTA ~= 0.74k
```
Comment (menduz) [inline test/snapshots/cubes.ts.crdt:566]: Performance impact

File: packages/@dcl/ecs/src/serialization/crdt/putComponent.ts [hunk 4]
```diff
@@ -14,17 +13,12 @@ export namespace PutComponentOperation {
...
-    timestamp: number,
-    componentDefinition: ComponentDefinition<unknown>,
-    buf: ByteBuffer
-  ) {
+  export function write(entity: Entity, timestamp: number, componentId: number, data: Uint8Array, buf: ByteBuffer) {
     // reserve the beginning
     const startMessageOffset = buf.incrementWriteOffset(CRDT_MESSAGE_HEADER_LENGTH + MESSAGE_HEADER_LENGTH)
 
     // write body
-    componentDefinition.writeToByteBuffer(entity, buf)
```
Comment (leanmendoza) [inline packages/@dcl/ecs/src/serialization/crdt/putComponent.ts:27]: This prevented to allocate a new Uint8Array to serialize the component and directly serializing in `buf`

Comment (leanmendoza) [reply packages/@dcl/ecs/src/serialization/crdt/putComponent.ts:27]: We can take the note and optimize in the future

**Review by leanmendoza — APPROVED** [review]

**Pull request merged**